### ISSUE TITLE
Feature request : deref takes an optional type

### ISSUE THREAD

**Issue by npadmana** [chapel-lang/chapel#11898]
When interoperating with C pointers, I find myself needing to cast from one type to another (most often when dealing with `c_void_ptr`), before a deref.  While this is straightforward to do now  eg:
`` var x  = (ptr_to_void : c_ptr(c_double)).deref()``, it might be nice to write this as ``var x = ptr_to_void.deref(c_double)``.

I believe this is a relatively simple overload to `deref` in `CPtr.chpl` (which I'm happy to file a PR for), so this is a question of whether this would be a desirable feature or not.

**Comment by bradcray** [comment]
I'm open to this proposal, though it's worth saying that I'm also not particularly invested in the current CPtr.chpl interface (so might not be the best person to ask).  The main reason I can think of not to support it might simply be one of purity—that the "right" way to dereference a c `void*` is to cast it to another pointer type first as in the workaround you give.

If we do add this, I think it should only be supported on the `c_void_ptr` type (which is what I'm guessing you're proposing as well?).  I.e., I don't think that `c_ptr`'s deref() method should take a type argument as well.

I'm guessing this is as simple as writing the following (?):

```chapel
proc c_void_ptr.deref(type t) {
  return (this: c_ptr(t)).deref();
}
```

If so, this also suggests that if others objected to it being included in CPtr.chpl, a program could easily write this as a convenience function, or we could add it to some other PtrUtils.chpl package module, or something...

**Comment by npadmana** [comment]
@bradcray -- I agree that the primary use case is `c_void_ptr`, but there are cases where one might want to reinterpret a pointer as something different. An immediate example is a real->complex in-place FFT where you reuse the array storage, but now treat it as an array of half as many complex numbers. Chapel already allows doing this; this would just make it a little less verbose.

That being said, maybe the method name should not be `deref`, but `derefAs` or something like that, to make it clear that you are doing a cast somewhere in the code.

**Comment by bradcray** [comment]
Using a different method name definitely makes me less averse to adding such a capability to vanilla c_ptrs.

I understand the general desire for using an array of complexes like an array of reals or vice-versa.  I understand less well why c_ptr's would be heavily involved in such operations (or, more generally, for anything much more than C interop itself).

**Comment by npadmana** [comment]
I agree - treating one array as another isn't a c_ptr issue (it's just that some of the cases are when I'm interoperating with C). I'll note that using recasting a c_ptr to a Chapel array into a different form and then wrapping it in a different Chapel array does get you this functionality (with the perils of memory layout etc).
